src/main/java/bio/terra/service/filedata/azure/util/BlobContainerClientFactory.java

```diff
@@ -0,0 +1,201 @@
+package bio.terra.service.filedata.azure.util;
+
+import com.azure.core.credential.AzureSasCredential;
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.http.HttpClient;
+import com.azure.storage.blob.*;
+import com.azure.storage.blob.models.UserDelegationKey;
+import com.azure.storage.blob.sas.BlobSasPermission;
+import com.azure.storage.blob.sas.BlobServiceSasSignatureValues;
+import com.azure.storage.common.StorageSharedKeyCredential;
+import com.azure.storage.common.sas.SasProtocol;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.Locale;
+import java.util.Objects;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * A class that wraps the SDK container client and facilities the creation of SAS Urls for blobs by
+ * abstracting the different strategies that depend on the authentication mechanism.
+ */
+public class BlobContainerClientFactory {
```

Comment:
This class name doesn't seem to match the class's function. From the name, I would expect that it is a factory that creates `BlobContainerClient`s. But it seems to be two things: a thin wrapper around a `BlobContainerClient`, and the provider of the `createReadOnlySasUrlForBlob()` API.

Specifically, it's not a factory because it only creates a single client during object construction.

Reply:
I've looked at this and the enum use more, and I think I better understand what's going on here.

There _is_ a factory operation here, where it can create one of three flavors of "wrappers" around a `BlobContainerClient`. Depending on which type you have, some fields will be null/unused, others will be used. I think the right way to model this is to keep the notion of a factory, but have it return one of three subclasses depending on which type of wrapper is needed. e.g.

```
abstract class BlobContainerClientWrapper {
  abstract public String createReadOnlySasUrlForBlob(String blobName);
  class SharedKeyContainerClient extends BlobContainerClientWrapper { ... }
  class UserDelegatedKeyContainerClient extends BlobContainerClientWrapper { ... }
  class SasTokenContainerClient extends BlobContainerClientWrapper { ... }
  static BlobContainerClientWrapper createClient(String accountName, String accountKey, String containerName) { ... new SharedKeyContainerClient() ... }
    [ ... etc ... ]
}
```
I don't like the name `wrapper`; maybe `Dao` or `PDao` would make more sense? But I think the idea will help clean up the code.

Reply:
I realize this suggestion isn't much to go on, so I wrote up a more complete version of what this would look like. There's still a bit of duplicate code, and the class name isn't great, but I think the subclass model better fits the code here.
[BlobContainerClientWrapper.java.txt](https://github.com/DataBiosphere/jade-data-repo/files/6818260/BlobContainerClientWrapper.java.txt)

Reply:
Thanks for looking into this with this level of detail. I appreciate it. You've pinpointed something I struggled with as well.   

I had initially implemented something very similar to the approach you suggested, and I found it a bit confusing to introduce wrappers to the same underlying object (BlobContainerClient) as a mean to determine the strategy to generate the SAS tokens - the border-line silly names I ended up with, were BlobContainerClientWrapperWithContainerSASUrlStrategy, BlobContainerClientWrapperWithBlobContainerClientStrategy and BlobContainerClientWapperWithUserDelegatedKeyStrategy, :).   

Having said all these and rethinking this again based on your feedback. I think a better approach is to create an abstraction for the SAS token generation logic . So an interface like this:

```java
public interface BlobSasUrlFactory {
  String createReadOnlySasUrlForBlob(String blobName);
}
```
An implementations for each strategy:

```java
public class ContainerSasTokenSasUrlFactory implements BlobSasUrlFactory

public class UserdelegatedKeySasUrlFactory implements BlobSasUrlFactory

public class SharedKeySasUrlFactory implements BlobSasUrlFactory
```
And depending of the constructor called, the properly implementation is called, e.g.:

>Note: I would rename `BlobContainerClientFactory` to `BlobContainerWrapper` (I also don't like the name wrapper, but I can't think of anything better)

```java
public class BlobContainerWrapper {

  private final BlobSasUrlFactory sasUrlFactory;

...
  public BlobContainerWrapper(String accountName, String accountKey, String containerName) {
    ...
    sasUrlFactory = new SharedKeySasUrlFactory(...);
 }


  public String createReadOnlySasUrlForBlob(String blobName) {
    return sasUrlFactory.createReadOnlySasUrlForBlob(blobName);
}

```

I think this approach will remove of 'switching' strategies using an `enum` which seems at the core of the issue/feedback.  Thoughts?

Reply:
Sorry, I was on vacation and didn't have a chance to reply to this comment. I haven't looked at where the code ended up, but I'm sure it will be OK. The code can always be revised in the future.

I hope that your next PR will smaller, or if you need to create a large PR, has a draft PR where discussion can begin earlier. Larger PRs are much harder to provide feedback on, and if you've written a lot of code and get a lot of feedback, can result in taking up more of your time than would otherwise be necessary.